Xamarin.Forms.Platform.iOS/Renderers/ListViewRenderer.cs

```diff
@@ -692,22 +695,56 @@ internal nfloat GetEstimatedRowHeight(UITableView table)
 				return CalculateHeightForCell(table, firstCell);
 			}
 
-			public override nfloat EstimatedHeight(UITableView tableView, NSIndexPath indexPath)
+			//public override nfloat EstimatedHeight(UITableView tableView, NSIndexPath indexPath)
+			//	=> 300;
+			//public override nfloat EstimatedHeight(UITableView tableView, NSIndexPath indexPath)
+			//{
+			//	if (_useEstimatedRowHeight)
+			//		return tableView.EstimatedRowHeight;
+
+			//	// Note: It is *not* an optimization to first check if the array has any values.
+			//	nfloat specifiedRowHeight;
+			//	if (_rowHeights.TryGetValue(indexPath, out specifiedRowHeight) && specifiedRowHeight > 0)
+			//		return specifiedRowHeight;
+
+			//	return UITableView.AutomaticDimension;
+			//}
+
+			internal override void InvalidatePrototypicalCellCache()
 			{
-				if (_useEstimatedRowHeight)
-					return tableView.EstimatedRowHeight;
+				_cellByIndex.Clear();
+				_cellByType.Clear();
+			}
+
+			internal Cell GetPrototypicalCell(NSIndexPath indexPath)
+			{
+				var cachingStrategy = List.CachingStrategy;
+				if ((cachingStrategy & ListViewCachingStrategy.RecycleElementAndDataTemplate) !=
+					ListViewCachingStrategy.RecycleElementAndDataTemplate)
+					return GetCellForPath(indexPath);
 
-				// Note: It is *not* an optimization to first check if the array has any values.
-				nfloat specifiedRowHeight;
-				if (_rowHeights.TryGetValue(indexPath, out specifiedRowHeight) && specifiedRowHeight > 0)
-					return specifiedRowHeight;
+				Cell protoCell;
+				if (!_cellByIndex.TryGetValue(indexPath, out protoCell))
+				{
+					var itemType = GetItemTypeForPath(indexPath);
+					if (!_cellByType.TryGetValue(itemType, out protoCell))
+					{
+						// cache prototypical cell by item type; Items of the same Type share
+						// the same DataTemplate (this is enforced by RecycleElementAndDataTemplate)
+						protoCell = GetCellForPath(indexPath);
+						_cellByType[itemType] = protoCell;
+					}
 
-				return UITableView.AutomaticDimension;
+					_cellByIndex[indexPath] = protoCell;
+				}
+
+				var templatedItems = GetTemplatedItemsListForPath(indexPath);
+				return templatedItems.UpdateContent(protoCell, indexPath.Row);
 			}
 
 			public override nfloat GetHeightForRow(UITableView tableView, NSIndexPath indexPath)
 			{
-				var cell = GetCellForPath(indexPath);
+				var cell = GetPrototypicalCell(indexPath);
```

Comment:
Where it all comes together! If a prototypical cell exists for the the type of item at the index path then use that instead of recreating a view for each height measurement.